### ISSUE TITLE
[Feature Request] The ability to resurrect NPC's at a specified town hall, regardless of upkeep location

### ISSUE THREAD

**Issue by jumpsplat120** [P3pp3rF1y/AncientWarfare2#369]
Currently it seems that NPC's are resurrected at their designated Town Hall or nearest(?) Town Hall. Thing is, if I have a Town Hall in a tavern where the NPC's are eating, I don't also want them to resurrect there, nor do I want a priest to have to be there. If possible, I'd like to designate a "main" Town Hall, so that I can have them resurrected at a church or something more thematically appropriate. Also that way if NPC Joe dies near the barracks Town Hall, NPC Bob dies near the tavern Town Hall and NPC Sheryl dies near the farm Town Hall, then they can all be resurrected at the church Town Hall, instead of having multiple priests for each one, or having the priest run from building to building.

**Comment by Sandriell** [comment]
I would love to see something like this implemented as well.

**Comment by GotoLink** [comment]
Putting some notes for later:
-routing orders for priests ?
-town hall upgrade for sharing the death lists ? (research need ?)
-church block to pay gold for resurrections ? (research need ?)

In any case, this is an interesting subject, as it would require quite work to reorganize the resurrection code.

**Comment by Dulciphi** [comment]
What would you think about an altar for the Priest? This would be a block dedicated to resurrection (much like your suggestion of a church block to pay gold for resurrections) - except NPCs would only resurrect there if it's being "worked" by a Priest when they die. Secondary to that - the altar could even give off an area of effect buff. Sort of like a beacon for NPCs. A Priest could only perform one task at a time. Either resurrecting or "working" the beacon. Resurrection should take a while - at least a minute in real time (configurable) and the "beacon" effects would be absent during this time.

I say this because, currently, Priests are a little (a lot) OP. I've had players keep their Priest in a chest for "when needed". On NPC death, the players would simply spawn their priest in and let him do his thing before packing him safely back away in item form. I'm looking for an incentive to NOT do that. 

I was thinking this would require an altar being in the "active" state -and within range- when an NPC dies - else they are gone for good ...but, thinking about your own "town hall upgrade" suggestion, I believe that could work even better. A sort of Town Hall "network"? It certainly allows for expansion.. and at this point I should probably go make a separate issue. See,  what I'm really after is a Town Hall block that's both beneficial and critical. I've made the town hall block a target for griefing. (Think along the lines of the Invasion mod, except this block actually has a point and a purpose other than to merely attract enemies). The remaining issue I face is that chests can currently replace town hall blocks in just about every way. I'd be after an option to disable ordinary chests being used as an upkeep point. 

Anyway, pardon my off-topic wanderings there. It's an interesting topic indeed.

**Comment by jumpsplat120** [comment]
I really like the concept of an altar for a Priest, which would tie in well with the leveling that Priests (and all NPC's) already have. The higher the level of a Priest, the wider the range of the altar's effects. The buff could be like a slight increase in work efficiency, something like 1.15x more efficient, non stackable with other altars, of course. This would also help with the keeping Priests out of the box as well, since they'd want their Priests to be the max level so that the altar reaches it's widest range. Resurrecting really is an OP concept, but only for OP NPC's. I think it shouldn't take a minute real time for each NPC, but x*10 seconds, where x is the level of the NPC. That way if a whole bunch of low level workers were killed in the beginning of your game, it wouldn't take ten minutes to bring them all back. If it did, people would just rather craft new ones.

A town hall network is an interesting concept, but really the problem here is that they are called "Town Halls". You only have one Town Hall in a town generally. What we need is an upkeep block that operates slightly differently, one where NPC's can get food only. A Town Hall should be a block that is worked by a NPC (think like a mayor or a king or lord) and have a separate function. The speculations for Town Blocks are very useful.

Town Block Ideas:
- It has a player name (or kingdom name for NPC run kingdoms) within the block. It must be worked by an NPC to operate, and if it the NPC is killed, the name can be changed, effectively "taking over" that kingdom. Then the new player can do with all of the NPC's of that kingdom as he pleases.
- Without the a NPC working the block, NPC's just wander around and don't work or do anything.
- There can be a main Town Hall block (It could be renamed the King's block for easy of understanding) and a Lordship block as well. The King's block is the main Town Hall block, and a lordship is an extension of that Town Hall. It operates the same except for one difference. NPC's are attached to the Lordship block and Lordship blocks are X distance away to a King's block, which links them. And then you can link lordships to lordships, giving you a way of extending your kingdom. If the NPC working a lordship is killed then all NPC's attached to it, and all blocks "lower" in the chain are also removed from control of the kingship. So each lordship farther away from the King's block is less valuable, meaning you would have to grow the kingdom organically to keep it "under control".
- To attach an NPC to a Lordship, you simply place it into the GUI of the Lordship, once you've got a name typed into the top of the block, designating it as yours. 

There's lots that you could do with it, but that's one way the network would work, and it definitely separates food and "government" as it were. 

Also, are you working on the mod as well?

**Comment by Dulciphi** [comment]
^yes. wait.. who, me? No, I mean no. lol I'm not a coder, unfortunately. I merely designed the banner and tested a bunch of stuff - and otherwise make a nuisance of myself here and there. That's a huge "yes" to your suggestions, though. That's very much along the lines of what i had in mind and discussed with Shadowmage at length. The issue he had was with generating 'raids' to attack and destroy the "King's block", as you're calling it here (I think I called them "Town Hall" and "outposts"). He was thinking in terms of "walls" and how nothing could prevent a player from merely building their walls out to Timbuktu, spamming the inside with torches and then living in safety - and that's fine. If someone can get their defences that solid, they deserve them. The idea was to give an **imperative** to build defences. Otherwise they're pointless. There's also plenty of game mechanics to handle spawns now. I have no trouble spawning mobs and adjusting their range to match the range of Town blocks. The problem I had was that players could merely use chests instead of Town Halls. I made it so that chests can be griefed as well, and that did help.. 

Tl;dr: Townhall blocks need the emphasis put back on them. I can handle griefing and spawns that require the blocks be well defended, but in order for that to be feasible, the blocks themselves need to be "needed".
